resize.js

```diff
@@ -15,218 +18,216 @@ var S3 = new AWS.S3();
+        if (!err && data) {
+            // It is in the cache, so redirect to there
+            log('info', 'cache hit for ' + fileName + '.' + fileType + '(' + resolutionX + 'x' + resolutionY + 'px)');
+            res.writeHead(302, {'Location': data.url, 'Cache-Control': 'public'});
+            res.end();
+            return;
+        }
         // It does not exist in the cache, so generate and upload
         res.writeHead(200, {'Content-Type': supportedFileType(fileType)});
         Image.encodeAndUpload(fileName, fileType, resolutionX, resolutionY, res);
-      }
     });
-}
-Image.encodeAndUpload = function(fileName, fileType, resolutionX, resolutionY, res ) {
-  file = config.get('originals_dir') + '/' + fileName;
-  // Get the image and resize it
-  gm(file)
-    .options({imageMagick: true})
-    .resize(resolutionX, resolutionY)
-    .stream(fileType, function streamOut(err, stdout, stderr) {
-      r = stdout.pipe(res);
-      r.on('finish',function() {
-        res.end();
-      });
-    });
-
-  gm(file)
-    .options({imageMagick: true})
-    .resize(resolutionX, resolutionY)
-    .toBuffer(fileType, function write(err,stream) {
-      if ( !err) {
-        // This might mean we have generated the same file while an upload was in progress.
-        // However this is still better than not being able to server the image
-        Image.uploadToCache(fileName, fileType, resolutionX, resolutionY, stream);
-      }
-    }); 
-}
-Image.uploadToCache = function(fileName, fileType, resolutionX, resolutionY, content) {
-  // Upload to AWS
-  key = fileName + '_' + resolutionX + 'x' + resolutionY + '.' + fileType;
-  upload_params = {
-    Bucket: config.get('aws.bucket'),
-    Key: key,
-    ACL: 'public-read',
-    Body: content,
-    // We let the client cache this for a month
-    Expires: (new Date()).setMonth(new Date().getMonth()+1)/1000,
-    ContentType: supportedFileType(fileType),
-    // We let any intermediate server cache this result as well
-    CacheControl: 'public'
-
-  };
-  S3.putObject(upload_params, function(err,data) {
-    if ( err ) {
-      log('error','AWS upload error: ' + JSON.stringify(err));
-    } else {
-      log('info','Uploading of ' + key + ' went very well');
-        url = config.get('aws.bucket_url') + '/' + key;
-        insertImage.run([fileName, resolutionX, resolutionY, supportedFileType(fileType), url],function(err) {
-          if ( err ) { console.error(err); }
+};
+Image.encodeAndUpload = function (fileName, fileType, resolutionX, resolutionY, res) {
+    var file = config.get('originals_dir') + '/' + fileName;
+    // Get the image and resize it
+    gm(file)
+        .options({imageMagick: true})
+        .resize(resolutionX, resolutionY)
+        .stream(fileType, function (err, stdout, stderr) {
+            var r = stdout.pipe(res);
+            r.on('finish', function () {
+                res.end();
+            });
```

Comment:
That was what I was talking about: `res` is just a writeable stream and will be closed by `stdout` when `stdout` closes. In fact, in your code `r === res`. So basically you're now closing `res` as soon as you get the event that `res` has been closed :wink: